wandb/sdk/launch/sweeps/scheduler.py

```diff
@@ -0,0 +1,206 @@
+            logger.debug(_msg)
+            wandb.termlog(_msg)
+            self.state = SchedulerState.CANCELLED
+            self.exit()
+            return
+        except Exception as e:
+            _msg = f"Scheduler failed with exception {e}"
+            logger.debug(_msg)
+            wandb.termlog(_msg)
+            self.state = SchedulerState.FAILED
+            self.exit()
+            raise e
+        else:
+            _msg = "Scheduler completed."
+            logger.debug(_msg)
+            wandb.termlog(_msg)
+            self.exit()
+
+    def exit(self) -> None:
+        self._exit()
+        if self.state not in [
+            SchedulerState.COMPLETED,
+            SchedulerState.CANCELLED,
+        ]:
+            self.state = SchedulerState.FAILED
+        for run_id in self._runs.keys():
+            self._stop_run(run_id)
+
+    def _update_run_states(self) -> None:
+        for run_id, run in self._runs.items():
+            try:
+                _state = self._api.get_run_state(self._entity, self._project, run_id)
+                if _state is None or _state in [
+                    "crashed",
+                    "failed",
+                    "killed",
+                    "finished",
+                ]:
+                    run.state = SimpleRunState.DEAD
+                elif _state in [
+                    "running",
+                    "pending",
+                    "preempted",
+                    "preempting",
+                ]:
+                    run.state = SimpleRunState.ALIVE
+            except Exception as e:
+                _msg = f"Issue when getting RunState for Run {run_id}: {e}"
+                logger.debug(_msg)
+                wandb.termlog(_msg)
+                run.state = SimpleRunState.UNKNOWN
+                continue
+
+    def _add_to_launch_queue(
+        self,
+        uri: Optional[str] = None,
+        resource: Optional[str] = None,
+        entry_point: Optional[List[str]] = None,
+        run_id: Optional[str] = None,
+    ) -> "public.QueuedJob":
+        """Add a launch job to the Launch RunQueue."""
+        run_id = run_id or generate_id()
+        job = launch_add(
+            uri or os.environ.get(wandb.env.DIR, os.getcwd()) or "",
+            project=self._project,
+            entity=self._entity,
+            queue=self._launch_queue,
+            resource=resource,
+            entry_point=entry_point,
+            run_id=run_id,
+        )
+        self._runs[run_id].launch_job = job
+        _msg = f"Added job to Launch RunQueue: {self._launch_queue} RunID:{run_id}."
+        logger.debug(_msg)
+        wandb.termlog(_msg)
+        return job
+
+    def _stop_run(self, run_id: str) -> None:
+        _msg = f"Stopping run {run_id}."
+        logger.debug(_msg)
+        wandb.termlog(_msg)
+        run = self._runs.get(run_id, None)
+        if run is not None:
+            run.state = SimpleRunState.DEAD
```

Comment:
is this actually how you stop a run lol, don't you need to make an API call to launch to kill it?

Reply:
The scheduler can only put the runs on the queue and then query their state via the backend. I think only the launch-agent has the handle to the actual processes to kill them: https://github.com/wandb/client/blob/225c28f6f62519b6acc2a93aa4fcc20d0ffec748/wandb/sdk/launch/agent/agent.py#L235.